### ISSUE TITLE
WireMock evaluation of all criteria generates unnecessary warnings

### ISSUE THREAD

**Issue by SebastianStavar** [tomakehurst/wiremock#776]
**[Description]** 
  _Short version:_ Non-json requests are evaluated against json mappings with body patterns and fill the log with failed json validation warnings.
  _Longer version:_ Because WireMock does an aggregation of the results for all evaluation rules (RequestPattern.match(Request request)) instead of hierarchically allowing a subsequent criterion to be evaluated against only after the higher one was matched, mappings that would immediately be excluded based on things like URL are still evaluated for body pattern matches. This is an issue when json body pattern matching mappings exists because it tries to evaluate ALL requests [that did not previously match] against the json body pattern matching mappings, which fills the log with all kind of json validation warnings.

**[Repro]** 

1. Start WireMock.
2. Train WireMock [use something like Postman and POST to WireMockServer:WireMockPort/__admin/mappings/] with RAW body:

`        {
            "id": "fab6d56b-d66e-4a5a-810a-ab323e485103",
            "request": {
                "urlPath": "/unavailablepath",
                "method": "POST",
                "headers": {
                    "X-SessionId": {
                        "equalTo": "90930D577144AB198B2082CA56E6A0F5",
                        "caseInsensitive": true
                    }
                },
                "bodyPatterns": [
                    {
                        "matchesJsonPath": "$.params[0][?(@.parameterone == 'something')]"
                    }
                ]
            },
            "response": {
                "status": 200,
                "bodyFileName": "a.json",
                "headers": {
                    "Content-Type": "application/json"
                }
            },
            "uuid": "fab6d56b-d66e-4a5a-810a-ab323e485103"
        }`

4. Make sure the _a.json_ file exists. I do not believe its content is relevant, but when I was creating the repro for this bug this is what I put in mine:

`{
  "jsonrpc": "2.0",
  "id": "-1412417872994864924",
  "result": {
    "payload": {
      "timeStamp": "2017-02-17T16:26:28.588"
    }
  }
}`

5. Send a non-admin request to WireMock. I used an empty GET to WireMockServer:WireMockPort/services/none
6. Observe the WireMock console. Result contains something like (actual error differs based on body patterns in mappings and on the requests):

> 20:41:12 INFO  WireMock - Warning: JSON path expression '$.params[0][?(@.parameterone == 'something')]' failed to match document '' because of error 'json string can not be null or empty'
20:41:12 INFO  WireMock - Warning: JSON path expression '$.params[0][?(@.parameterone == 'something')]' failed to match document '' because of error 'json string can not be null or empty'

The number of such entries considerably increases when the mappings are more complex and the requests are quite large. I think it is a one to one matching for each line of the request to each body pattern and each will generate at least a warning. 

This makes our logs (which are already quite large) extremely large and it is very hard to follow what is wrong when you need to navigate through hundreds of pages of such warnings.

We also do not want to downgrade the logging level since we would like to be able to observe the [other] information provided.

**[Environment]**

We are using WireMock 2.5.1, but we tested (before actually debugging the code and finding the root cause) and the issue is still present in WireMock 2.8.0.

**Comment by AlixBa** [comment]
related (no search on the root cause): 
https://groups.google.com/forum/#!topic/wiremock-user/DOmiAQ83W8E

**Comment by jadzya** [comment]
TLDR: This bug brings wiremock to its knees and causes multiple and resonably random failures that are very hard to investigate. This is no longer just about the size and quality of the logs, it majorly affects performance.

Long version: This is a lot worse than we initially estimated. We are running 160 simultaneous tests (with about 5-15 requests per test) and we were having a failure rate of 90%. It was really hard to figure out what the problem was since wiremock seemed properly trained and working so we were actually adding more debug messages to our app in trying to figure out where the problem was. But we did the right thing and decided to fix the wiremock logging issue first so we can actually see the errors in the much smaller/cleaner logs once the issue was removed . As soon as the errors stopped being logged our failure rate moved to 3% and we could easily figure out what the causes were for those.

**Comment by philippe-granet** [comment]
See my comment here https://github.com/tomakehurst/wiremock/issues/652#issuecomment-446423920

**Comment by philippe-granet** [comment]
See also my PR https://github.com/tomakehurst/wiremock/pull/1049

**Comment by tomakehurst** [comment]
Closing individual logging issues in favour of (referenced) logging redesign issue.
